search: Single enter should narrow to current filters.

pratik-pc opened the pull request:
Currently, when we remove a search filter pill, we need to click
enter twice (once to update the pill from suggestion and again to
update narrow state) to navigate to the appropriate filter narrow.

This changes the behavior to:
If the search query has focus but no typeahead item is selected or highlighted, we navigate on Enter. We also keep the search bar open on navigation to match the behavior from when we add pills.

Pratik:
> So basically:
> 1) keyboard deletion moves focus to the input field instead of the pill and keeps the typeahead open
> 2) mouse deletion moves focus to the input field but doesn’t open the typeahead.
And we trigger the search on Enter in both cases?

Alya:
> Yeah, I think we can try that.

Intended behavior discussed in [#issues > 📂 Search filtering does not work seamlessly @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2441516)

Fixes: zulip#33416.

|Before|After|
|-|-|
|<video src="https://github.com/user-attachments/assets/b5b2f8f4-b83f-4f87-83d5-453e95ba391b" />|<video src="https://github.com/user-attachments/assets/bb188283-2ce5-45a4-ad01-9e6cf18e0e51" />|
|Clicking on `x` opens typeahead|Clicking on `x` doesn't open typeahead|
|<video src="https://github.com/user-attachments/assets/3a9dc0dc-88be-4790-8fcd-aa078f88e221"  />|<video src="https://github.com/user-attachments/assets/e88f06fd-0888-486e-9d1c-dc93b8a636a8" />|
|Narrow on `Enter` triggers `blur`|Brings back focus to search on `Enter`|
|<video src="https://github.com/user-attachments/assets/3ace40eb-51a0-427b-8c02-3864710796ab" />|<video src="https://github.com/user-attachments/assets/fa1b91fe-76b3-4832-9dd2-da366d260da2" />|


<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

ligmitz commented:
Looks good @pratik-pc !

evykassirer commented on web/src/search.ts:
```diff
@@ -125,10 +128,7 @@ function narrow_to_search_contents_with_search_bar_open(): void {
-    // The shake animation will show if there is any invalid term in the,
-    // search bar, even if it's not what the user just typed or selected.
     if (!text_terms.every((term) => Filter.is_valid_search_term(term))) {
-        $("#search_query").addClass("shake");
```
If you're removing the shake here, I'd rename this function to `text_terms_are_valid()` or something like that

pratik-pc replied:
I think this was from an earlier outdated version and wasn't present in the previous version.

evykassirer commented:
Single enter already updates the narrow to the search filters on CZO, as far as I can tell. Am I missing something?

evykassirer commented:
That should be in the PR description and commit message to make that clearer.

But looking more into this now, https://github.com/zulip/zulip/issues/33416 says that the fix should be updating the narrow immediately when removing a pill, which I think is the right move and is a different solution than the one you've done here.

evykassirer commented:
Ah yeah, I see. I'm pretty sure we want removing pills to work the same as adding pills, where the narrow gets updated right away.

pratik-pc committed:
search: Narrow on enter when $search_query_box has focus.

The Enter handler suppressed narrowing whenever the typeahead was
merely open, on the assumption that Enter was selecting a typeahead
item. That assumption no longer holds: the typeahead can be open with
nothing highlighted (for an empty query, right after a pill is removed
or using up arrow to search box), and in that state Enter should narrow
to the current search terms rather than do nothing.

We now distinguish "the typeahead was open" from "an item was selected
from the typeahead" using a new has_active_item() helper, and only
suppress narrowing in the latter case. The flag is renamed from
typeahead_was_open_on_enter to typeahead_item_was_selected_on_enter
to reflect this.

Fixes #33416.

Discussion: https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2441516

pratik-pc committed:
search: Focus input after removing a search pill via backspace.

Pressing backspace on a focused search pill removed it but then
moved focus to the preceding pill. That is input_pill's default
behavior, and it is intentional: it lets users chain-delete pills
by holding backspace.

For search, the more useful behavior is to return focus to the
input, so the user can immediately keep editing or typing their
query. This matches how removing a pill via its `x` button or the
expand control already behaves.

We add a `focus_input_on_backspace_remove` option to input_pill,
defaulting to off so behavior elsewhere is unchanged, and enable it
for search pills.

Fixes part of #33416.

Discussion: https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2441516

pratik-pc committed:
search: Don't open typeahead when a search pill is removed.

Removing a search pill triggers a "change" event on the search box
via `removePill`, and our "change" handler unconditionally re-ran
the typeahead lookup. As a result, clicking a pill's `x` to remove
it would pop open the typeahead, even though the user was not
editing the query.

We now re-run the lookup only when the typeahead is already visible,
so removing a pill refreshes an open typeahead but no longer opens a
closed one.

Fixes part of #33416.

Discussion: https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2441516

evykassirer commented on web/src/search.ts:
```diff
@@ -144,7 +144,6 @@ function narrow_to_search_contents_with_search_bar_open(): void {
     if ($(".navbar-search.expanded").length === 0) {
         open_search_bar_and_close_narrow_description();
         focus_search_input_at_end();
-        search_typeahead.lookup(false);
```
The commit message says this was "redundant", which suggests removing it wouldn't cause a functional change, but it seems like this is a functional change, something more like `search: Don't reopen typeahead after narrowing from a selected pill.`.

pratik-pc replied:
This got fixed in a separate PR.

evykassirer commented on web/src/input_pill.ts:
```diff
@@ -457,12 +457,12 @@ export function create<ItemType extends {type: string}>(
                     } else {
                         $next.trigger("focus");
                     }
+                    funcs.removePill(util.the($pill), "backspace");
```
This reorder lives in `input_pill.ts`, which is shared by ~15 widgets, not just search. End-state focus should be unchanged for other consumers (none of their `onPillRemove` callbacks move focus), but `$prev` now briefly gets focus before the deleted pill is removed from the DOM, so it seems worth confirming you've manually tested backspace pill removal in compose recipients and a settings picker.

Also worth making sure the prev/next getting temporarily focused doesn't cause any issues (e.g styling flash, screen reader announcement)
                                                                                                                                    
> We also reorder the removePill call in input_pill to happen after
the focus assignment, so that the onPillRemove callback fires while
the adjacent pill elements are still in their expected positions.

Also, it doesn't seem like that's the reason here? It seems more about focus ordering, not pill DOM positions. Or is there something more about position here that you can explain more?

pratik-pc replied:
I added a separate option to handle `backspace` removal, but I'm not sure whether we want to add more options. Calling `removePill()` after `trigger()` is also an option, as above, but it might trigger unnecessary focus jumps and focus handlers, either now or in the future.

evykassirer commented on web/src/search.ts:
```diff
@@ -304,7 +302,10 @@ export function initialize(opts: {on_narrow_search: OnNarrowSearch}): void {
+            typeahead_item_was_selected_on_enter =
+                keydown_util.is_enter_event(e) &&
+                search_typeahead.shown &&
+                search_typeahead.$menu.find(".active").length > 0;
```
It might be nice to add a public method in `bootstrap_typeahead` to make this more readable and hide the more messy way of having to check this:

```js
has_active_item(): boolean {
      return this.$menu.find(".active").length > 0;
  }
```

pratik-pc replied:
Done.

evykassirer commented on web/src/search.ts:
```diff
@@ -84,13 +84,11 @@ function narrow_or_search_for_term({on_narrow_search}: {on_narrow_search: OnNarr
-    // Narrowing will have already put some terms in the search box,
-    // so leave the current text in.
-    $("#search_query").trigger("blur");
+    if ($(".navbar-search.expanded").length === 0) {
```
We now do this in two places, so maybe we can make a helper function?

e.g.

```js
function reopen_search_bar_if_collapsed() {
    if ($(".navbar-search.expanded").length === 0) {
        open_search_bar_and_close_narrow_description();
        focus_search_input_at_end();
        search_input_has_changed = true;
    }
}
```

and then

```js
on_narrow_search(terms, {trigger: "search"});

// Reopen the search bar if the narrow collapsed it (which happens for common narrows)
reopen_search_bar_if_collapsed();
```

pratik-pc replied:
Done.

evykassirer reviewed (commented):
Looks good! I think this is all functionally correct, just wanted to clean up some of the descriptions of what's going on. Thanks for working on this, in such a confusing part of the codebase!

evykassirer commented:
Please reply to my comments with the changes you made and let me know when it's ready for another review :)

pratik-pc commented:
> Please reply to my comments with the changes you made and let me know when it's ready for another review :)

It is not ready for review yet. I’m waiting on #39211 before making some changes.

pratik-pc committed:
search: Keep search bar open and focused after narrowing on Enter.

When the user pressed Enter to narrow to the current search terms,
`narrow_or_search_for_term` blurred the search box. This left focus
outside the search bar and, for common narrows, caused the search bar
to close.

We instead keep the search bar open and focused after narrowing,
matching the behavior of the typeahead-selection path
(`narrow_to_search_contents_with_search_bar_open`). Both paths now
share a `reopen_search_bar_if_collapsed` helper, which reopens the
bar only when a common narrow collapsed it.

Fixes part of #33416.

Discussion: https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2441516

pratik-pc commented:
@evykassirer Addressed above feedbacks

evykassirer commented:
Asking about [#issues > 📂 Search filtering does not work seamlessly @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20Search.20filtering.20does.20not.20work.20seamlessly/near/2484246), sorry for not thinking about that earlier.

evykassirer commented on web/src/search.ts:
```diff
@@ -319,13 +322,11 @@ export function initialize(opts: {on_narrow_search: OnNarrowSearch}): void {
                 keydown_util.is_enter_event(e) &&
                 $search_query_box.is(":focus") &&
-                !typeahead_was_open_on_enter
+                !typeahead_item_was_selected_on_enter
```
It would be nice to still have a comment here.  Maybe something like this?

```js
 } else if (
      keydown_util.is_enter_event(e) &&
      $search_query_box.is(":focus") &&
      // When an item is highlighted, this Enter selected a typeahead
      // item and the typeahead's own handler narrows, so we must not
      // narrow again here.
      !typeahead_item_was_selected_on_enter
  ) {
```

pratik-pc commented:
Added the comment. I'll update the focus behavior for `#search_query` once we reach a consensus on the desired behavior.

pratik-pc committed:
commit to test input focus on pill removal.
